grails-app/services/org/pih/warehouse/inventory/CycleCountProductAvailabilityService.groovy

```diff
@@ -0,0 +1,139 @@
+ */
+@Transactional
+class CycleCountProductAvailabilityService {
+
+    ProductAvailabilityService productAvailabilityService
+
+    /**
+     * Get product availability for a given facility and product. Filter out items with QOH == 0.
+     */
+    List<AvailableItem> getAvailableItems(Location facility, Product product) {
+        return getAvailableItems(facility, [product])
+    }
+
+    /**
+     * Get product availability for a given facility and list of products. Filter out items with QOH == 0.
+     */
+    List<AvailableItem> getAvailableItems(Location facility, List<Product> products) {
+        List<String> productIds = products.collect { it.id }
+        return productAvailabilityService.getAvailableItems(facility, productIds, false, true)
+    }
+
+    /**
+     * Refreshes a given cycle count by fetching up-to-date product availability for the products associated with
+     * the count and updating the QoH for each of the items in the most recent count.
+     *
+     * @return true if one or more cycle count items were added to the count or had their QoH modified.
+     */
+    boolean refreshProductAvailability(CycleCount cycleCount) {
+        boolean itemsHaveChanged = false
+
+        // We only want to refresh the items of the most recent count. We do this because previous counts are already
+        // completed, so their data should be considered static.
+        Set<CycleCountItem> cycleCountItems = cycleCount.getItemsOfMostRecentCount()
+        int currentCountIndex = cycleCountItems.first().countIndex
+        Location facility = cycleCount.facility
+
+        // Used by newly created items. We know that the frontend only supports a single assignee and date counted
+        // for all items of a count, so for convenience we set the values of any new items to match.
+        User assigneeForNewItems = cycleCountItems.find{ it.assignee }?.assignee
+        Date dateCountedForNewItems = cycleCountItems.find{ it.dateCounted }?.dateCounted ?: new Date()
+
+        // This logic is shared between count and recount flows so we need to check which we're in.
+        CycleCountItemStatus statusForNewItems = cycleCount.status.isCounting() ?
+                CycleCountItemStatus.COUNTING : CycleCountItemStatus.INVESTIGATING
+
+        // The refresh is based on the current product availability in all [bin location + lot] combinations for all
+        // the products of the count. We do this because new items might have been created since the count started.
+        List<AvailableItem> availableItems = getAvailableItems(facility, cycleCount.products)
+        for (AvailableItem availableItem : availableItems) {
+            boolean itemHasChanged = updateItem(cycleCount, cycleCountItems, availableItem, facility, currentCountIndex,
+                    assigneeForNewItems, dateCountedForNewItems, statusForNewItems)
+
+            itemsHaveChanged = itemsHaveChanged || itemHasChanged
```

Comment:
I would add a comment explaining what is happening here. I had to think a longer time what it is doing.
I assume if the first item is changed, the flag is set to `true`, and then we don't even have to check further items in terms of `itemHasChanged`.